### ISSUE TITLE
Guidance on managing BERTopic models

### ISSUE THREAD

**Issue by TheAIMagics** [MaartenGr/BERTopic#1861]
I'm seeking some advice on managing BERTopic models for efficient topic clustering.

I've utilised BERTopic to cluster approximately 13,000 data which is three months data, resulting in around 130 distinct topics. To further reduce the number of topics, I implemented hierarchical topic modelling technique, effectively reducing the number of topics to 100.

Upon thorough analysis of both the topics and the associated documents, I've identified the necessity for additional topic merging. As a result, I manually merged several topics to refine the clustering outcome.

Now, I'm at a stage where I need to preserve this final model for future use. Specifically, I aim to employ it for predicting the topics of upcoming month's data.

I would greatly appreciate any insights or suggestions.

**Comment by MaartenGr** [comment]
Thanks for sharing your question. Your answer can be found in the [best practices section](https://maartengr.github.io/BERTopic/getting_started/best_practices/best_practices.html#serialization) that also details serialization. Generally, the most stable method of saving and using your model is through `safetensors` or `pytorch`. Using pickle can be very difficult to version control and the other methods are a bit more lenient when it comes to version control. Note that these methods do not use the underlying dimensionality reduction and clustering algorithms but instead directly calculate assignment based on the cosine similarity between document and topic embeddings.

**Comment by TheAIMagics** [comment]
@MaartenGr Thank you for your response. I have grasped the concept of utilizing various serialization methods to save the final BERTopic model efficiently. However, I encountered a challenge specific to my scenario. Along with merging several topics, I also modified the mappings of documents. How can I effectively handle this scenario while ensuring the integrity of the final model?

**Comment by MaartenGr** [comment]
> How can I effectively handle this scenario while ensuring the integrity of the final model?

I'm not quite sure what you mean. What issue are you currently facing?

**Comment by TheAIMagics** [comment]
Hi Maarten,

Thank you for your email.
Clarifying my use case: I ran BERTopic clustering for about 75000 documents
(that represents 3 months of data). Since some clusters were similar, I
merged them using ".merge_topics".

After this, I assessed the distribution of topics and associated documents
using:
topic_model.get_document_info(docs) Upon analysis of this distribution, I
identified certain documents that were not correctly assigned to the right
clusters. To rectify this, I manually relocated these documents to their
correct clusters.

Now when next month's data comes in, I want to predict the clusters and I
want the clustering to be done taking into account my manual updates (so
that I don't have to keep repeating the same manual exercise every month).
In other words, how do I effectively communicate the changes I am making
manually, to the model? (especially concerning the modifications made at
the document level.)

Thanks again for your help.

**Comment by MaartenGr** [comment]
> Upon analysis of this distribution, identified certain documents that were not correctly assigned to the right clusters. To rectify this, I manually relocated these documents to their correct clusters.

Assuming you did this manual relocation using `.update_topics`, the resulting topic embeddings should be updated to incorporate those changes. So if you have saved and loaded the model using `safetensors` or `pytorch`, these topic embeddings will be used to perform the topic assignments.

**Comment by TheAIMagics** [comment]
How can I use update_topics to facilitate the relocation of documents within clusters while ensuring the accuracy of the model's topic assignments?

**Comment by MaartenGr** [comment]
`.update_topics` recreates the topic embeddings based on the newly created topics, so the underlying assignment will also be updated assuming you use `safetensors` or `pytorch`.

**Comment by AILearnerMode** [comment]
Hi Maarten,

I have a similar question to TheAIMagics and still not able to wrap my head around it.  Putting together a dumbed down version of what I am looking for:

After training my docs, I get 2 clusters:
ClusterA docs: Damaged products, used products, dirty products, incorrect product
ClusterB docs: Wrong product received, wrong delivery, wrong item sent

Now I review these clusters and want to tell the model that "incorrect product" actually belongs to ClusterB. How do I do this programmatically so when I send my next set of data to this model for predictions,  anything related to "incorrect products" are correctly added to Cluster B.

Im not exactly sure how to do that with update_topics as this method seems to only update the topic representation, which may not include "incorrect product" at all.

**Comment by MaartenGr** [comment]
@AILearnerMode If you check the source code of `.update_topics` you will notice that the topic embeddings are also updated and not only the topic representation. Therefore, and assuming you are using the `safetensors/pytorch` method I described above, it should also update the assignment.

**Comment by TheAIMagics** [comment]
@MaartenGr 

1. Using update_topics method, I modified both the topics and the mappings of documents.
2. Saved the updated model using the safetensors 
3. Loaded the model for prediction tasks.

Despite following this workflow, the predictions generated by the loaded model remain consistent with the original model, rather than reflecting the modifications made to the topics and document mappings.
Code
```
topic_model.update_topics(docs=modified_docs, topics=modified_topics)

directory_path = "directory_path"
embedding_model = "sentence-transformers/all-MiniLM-L6-v2"
topic_model.save(directory_path, serialization="safetensors", save_ctfidf=True, save_embedding_model=embedding_model)

updated_model = BERTopic.load(directory_path)
new_docs = ["But app shows return failed."]
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = embedding_model.encode(new_docs, show_progress_bar=True)
new_topics, new_probs = updated_model.transform(new_docs, embeddings)
```

Any suggestions on how to resolve this issue?

**Comment by MaartenGr** [comment]
@TheAIMagics Could share the steps **before** `.update_topics` as it might relate to choices being made in `.update_topics`.

**Comment by MaartenGr** [comment]
Hmmmm, it does seem like the topic embeddings are not properly updated... Could you perhaps run the following before doing the inference:

```python
updated_model = BERTopic.load(directory_path, embedding_model="all-MiniLM-L6-v2")
updated_model._create_topic_vectors()
```

It should update the topic vectors using the embedding model.

**Comment by TheAIMagics** [comment]
@MaartenGr Thank you for your response

Model performs good in predicting relocated documents, but it fails when dealing with outliers, assigning them to random clusters instead of correctly identifying them as outliers (Topic -1).

Steps Taken:

To address this issue, I attempted to utilize the hdbscan_model, which employs an approximation method to predict new points. This approach successfully predicts outliers, but it overlooks the accurate prediction of relocated documents.
```
topic_model.update_topics(docs = modified_docs,topics = modified_topics)
topic_model._create_topic_vectors()

new_doc = ["I love apple fruit juice"]
embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = embedding_model.encode(new_docs, show_progress_bar=True)
topic_model.transform(new_doc, embeddings)
```

Challenge:

This approach does not handle the prediction of relocated documents properly. When using the hdbscan_model, the predictions generated by the loaded model remain consistent with the original model, which is not ideal.

I am open to saving the model using the pickle method for predictions. This might provide a workaround to ensure accurate predictions for both relocated documents and outliers. Any  suggestions on how to tackle this issue would be greatly appreciated.

**Comment by MaartenGr** [comment]
It makes sense that the method you tried does not work since the topic vectors are not taken into account when running the predictions for HDBSCAN. I suggested doing it after loading the model using either `pytorch` or `safetensors` since the topic vectors will be used in those cases. 

I believe manually changing the assignment can only be done by removing HDBSCAN since that is a clustering model that learns certain clusters. What you suggest is adapting HDBSCAN, which is quite difficult. 

Instead, I would still advise using the method I proposed earlier and see if you can identify a threshold for when something does or does not relate to an outlier. My expectation is that if it does not assign it to an outlier, you can still use the probabilities (by setting `topic_model.calculate_probabilities_ = True` after loading your model) to do a manual assignment. This manual assignment would consist of three sequential steps:
* If the largest probability is the outlier topic, assign the document to the outlier topic
* elIf the largest probability does not exceed a certain threshold, assign it to the outlier topic
* else, assign to the non-outlier topics with the largest probability
